source/Halibut/Transport/ConnectionManagerAsync.cs

```diff
@@ -0,0 +1,300 @@
+                    {
+                        connections.Remove(connection);
+                    }
+                }
+            }
+        }
+
+        public void ClearPooledConnections(ServiceEndPoint serviceEndPoint, ILog log)
+        {
+            using (connectionsLock.Lock())
+            {
+                pool.Clear(serviceEndPoint, log);
+            }
+        }
+
+        public async Task ClearPooledConnectionsAsync(ServiceEndPoint serviceEndPoint, ILog log, CancellationToken cancellationToken)
+        {
+            using (await connectionsLock.LockAsync(cancellationToken))
+            {
+                await pool.ClearAsync(serviceEndPoint, log, cancellationToken);
+            }
+        }
+        
+        public IReadOnlyCollection<IConnection> GetActiveConnections(ServiceEndPoint serviceEndPoint)
+        {
+            using (connectionsLock.Lock())
+            {
+                using (activeConnectionsLock.Lock())
+                {
+                    if (activeConnections.TryGetValue(serviceEndPoint, out var value))
+                    {
+                        return value.ToArray();
+                    }
+                }
+            }
+
+            return Array.Empty<IConnection>();
+        }
+
+        public void Disconnect(ServiceEndPoint serviceEndPoint, ILog log)
+        {
+            using (connectionsLock.Lock())
+            {
+                pool.Clear(serviceEndPoint, log);
+                var toDelete = new List<IConnection>();
+                using (activeConnectionsLock.Lock())
+                {
+                    if (activeConnections.TryGetValue(serviceEndPoint, out var activeConnectionsForEndpoint))
+                    {
+                        toDelete = activeConnectionsForEndpoint.ToList();
+                    }
+                }
+
+                foreach (var connection in toDelete)
+                {
+                    SafelyDisposeConnection(connection, log);
+                }
+            }
+        }
+
+        public async Task DisconnectAsync(ServiceEndPoint serviceEndPoint, ILog log, CancellationToken cancellationToken)
+        {
+            using (await connectionsLock.LockAsync(cancellationToken))
+            {
+                await pool.ClearAsync(serviceEndPoint, log, cancellationToken);
+                var toDelete = new List<IConnection>();
+                using (await activeConnectionsLock.LockAsync(cancellationToken))
+                {
+                    if (activeConnections.TryGetValue(serviceEndPoint, out var activeConnectionsForEndpoint))
+                    {
+                        toDelete = activeConnectionsForEndpoint.ToList();
+                    }
+                }
+
+                foreach (var connection in toDelete)
+                {
+                    SafelyDisposeConnection(connection, log);
+                }
+            }
+        }
+        
+        void OnConnectionDisposed(IConnection connection)
```

Comment:
Is an async version of this coming? could it be marked as TODO?

Reply:
Lol, the async version is already written in the next branch 😁  But I'll add a todo to be safe.